Add fork version to topic (#1624)
ethereum/eth2.0-specs: specs/phase0/p2p-interface.md

### Topics and messages

Topics are plain UTF-8 strings and are encoded on the wire as determined by protobuf (gossipsub messages are enveloped in protobuf messages). Topic strings have form: `/eth2/ForkVersion/Name/Encoding`. This defines both the type of data being sent on the topic and how the data field of the message is encoded.

> - `ForkVersion` - the hex-encoded bytes from `state.fork.current_version` of the head state of the client, as also seen in `Status.head_fork_version`.

Reviewer: Might be worth a note on guidance on gossip subnet lookahead wrt upcoming forks. Maybe a good note to add in the FAQs

Something like: How are gossipsub topics expected to be handled during forks?

> When a node is preparing for upcoming tasks (e.g. validator duty lookahead) on a gossipsub topic, the node should join the topic with the `ForkVersion` of the future epoch in which the task is to occur.

Author: we kind of had one that could be adapted

> - `ForkVersion` - the hex-encoded bytes from `state.fork.current_version` of the head state of the client, as also seen in `Status.head_fork_version`.

Reviewer: Might make sense to rephrase this in terms of fork version of the epoch of the relevant message, rather than that of the head state.

Because we might prep for and listen to multiple fork version topics of the same message type at once, it isn't _strictly_ related to the head state.

> - `ForkVersion` - the hex-encoded bytes from `state.fork.current_version` of the head state of the client, as also seen in `Status.head_fork_version`.

Reviewer: Maybe something like:
```suggestion
- `ForkVersion` - the hex-encoded bytes of the `fork_version` of the epoch of the related messages to be sent on the topic
```

- `Name` - see table below
- `Encoding` - the encoding strategy describes a specific representation of bytes that will be transmitted over the wire. See the [Encodings](#Encoding-strategies) section for further details.

The fork version is hex-encoded using the following scheme:
```python

>   ForkVersion = ''.join('{:02x}'.format(x) for x in state.fork.current_version)

Reviewer: ```suggestion
  ForkVersion = ''.join('{:02x}'.format(x) for x in fork_version)
```

```

> For example, the fork version `[0, 1, 2, 10]` will be encoded as `0001020a`.

Reviewer: Recommend we use the format/type as in the rest of the specs

```suggestion
For example, the fork version `Version('0x0001020a')` will be encoded as `0001020a`.
```

The payload is carried in the `data` field of a gossipsub message, and varies depending on the topic:

> | Name                                           | Message Type            |

Reviewer 2: It seems we are still using `TopicName` to describe the field in the following. (Section `Global topics
` and `Attestation subnets`)

Reviewer: updated this in #1652
